### PR TITLE
Task 3

### PR DESCRIPTION
**Pull request by Ania911**
Done, please review it )

### CODE REVIEW

File: config.yml [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  baseURL: http://localhost:8080 # Base URL for your REST API endpoints
+  authentication:
+    username: default
+    password: 1q2w3e
+
+# Test Environment Configuration
+environment:
+  envName: QA # Name of the test environment
+  appUrl: http://localhost:8080 # URL of the application under test
+  loginPage: http://localhost:8080/ui/#login?
```
Comment (mkovsinv) [inline config.yml]: rm   loginPage: http://localhost:8080/ui/#login?

File: src/main/java/core/assertion/ApiAssertions.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+    public static void assertResponseBodyContains(Response response, String expectedString) {
+        String responseBody = response.getBody().asString();
+        Assert.assertTrue(responseBody.contains(expectedString), "Response body does not contain the expected string: " + expectedString);
+    }
+
+    public static void assertJsonFieldInResponseBody(Response response, String fieldName, String expectedValue) {
+        JSONObject responseBodyJson = new JSONObject(response.getBody().asString());
+        Assert.assertEquals(responseBodyJson.getString(fieldName), expectedValue, "Value of field '" + fieldName + "' in response body does not match the expected value: " + expectedValue);
+    }
```
Comment (mkovsinv) [inline src/main/java/core/assertion/ApiAssertions.java:39]: String.format("[%s] Test failed [%s]", asd, asd1);

File: src/main/java/core/assertion/Assertion.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+import org.testng.Assert;
+
+import java.time.Duration;
+
+public class Assertion {
+    private static WebDriver driver;
+
+    public Assertion(WebDriver driver) {
+        this.driver = driver;
+    }
```
Comment (mkovsinv) [inline src/main/java/core/assertion/Assertion.java:36]: move WebDriver to page object & create action

File: src/main/java/core/configuration/ApiConfig.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
+package core.configuration;
+
+
+public class ApiConfig {
+
+    public static final String baseURI = "http://localhost:8080";
```
Comment (mkovsinv) [inline src/main/java/core/configuration/ApiConfig.java]: public to private

File: src/main/java/core/configuration/ConfigReader.java [hunk 5]
```diff
@@ -0,0 +1,18 @@
+package core.configuration;
+import org.yaml.snakeyaml.Yaml;
+import java.io.InputStream;
+import java.util.Map;
+
+public class ConfigReader {
+    private Map<String, Object> config;
+
+    public ConfigReader(String configFile) {
```
Comment (mkovsinv) [inline src/main/java/core/configuration/ConfigReader.java]: to singleton

File: src/main/java/core/configuration/Configs.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
+package core.configuration;
+
+import java.util.Map;
+
+public class Configs {
+    private static final ConfigReader configReader = new ConfigReader("config.yml");
+    private static final Map<String, Object> config = configReader.getConfig();
+
+    // Access configuration values
```
Comment (mkovsinv) [inline src/main/java/core/configuration/Configs.java]: to methods

File: src/main/java/core/pageActions/DashBoardActions.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
+package core.pageActions;
+
+import core.uttility.Log;
+import org.apache.commons.lang.RandomStringUtils;
+import pages.DashboardPage;
+import pages.SettingsPage;
+
+public class DashBoardActions {
+    private DashboardPage dashboardPage;
```
Comment (mkovsinv) [inline src/main/java/core/pageActions/DashBoardActions.java]: = new DashboardPage()

File: src/main/java/core/uttility/CommonUIActions.java [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+import org.openqa.selenium.By;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.ui.ExpectedConditions;
+import org.openqa.selenium.support.ui.WebDriverWait;
+
+import java.time.Duration;
+
+public class CommonUIActions {
```
Comment (mkovsinv) [inline src/main/java/core/uttility/CommonUIActions.java]: move to pages
BasePage
extend other pages from BasePage

File: src/main/java/pages/SettingsPage.java [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+        locators = new PageLocators();
+        commonActions = new CommonUIActions(driver);
+        driver.get(Configs.settingsPage);
+    }
+
+    public void openDropDown() {
+        commonActions.clickButton(locators.openDropdown);
+    }
+
+    public void dropDownOptions() {
```
Comment (mkovsinv) [inline src/main/java/pages/SettingsPage.java]: rename, add more details

File: src/main/java/locators/PageLocators.java [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+    // Setting Page
+    public By settingsButton = By.xpath("//span[@class = 'sidebarButton__btn-title-mobile--j8jhQ']");
+    public By submitButton = By.xpath("//button[@type= 'submit']");
+    public By openDropdown = By.xpath("(//span[@class=\"inputDropdown__value--2gB2s\"])[2]");
+    public By titles = By.xpath("//div[@title= '3 hours'] ");
+    public By successLogInMessage = By.xpath("//div[contains(@class, 'notificationItem__success')]//p[contains(text(), 'Signed in successfully')]\n");
+    public By successSettingUpdateMessage = By.xpath("//div[contains(@class, 'notificationItem__success')]//p[contains(text(), 'Project settings were successfully updated')]");
+
+    // Dashboard Page
+    public By openDashboard = By.xpath(" //a[contains(@class, 'sidebarButton__nav-link--2TC0L') and contains(@href, '#default_personal/dashboard')]");
```
Comment (mkovsinv) [inline src/main/java/locators/PageLocators.java]: default_personal replace

File: src/main/java/locators/PageLocators.java [hunk 11]
```diff
@@ -0,0 +1,32 @@
+package locators;
+
+import org.openqa.selenium.By;
+
+public class PageLocators {
+    // Log in page
+    public By userName = By.name("login");
+    public By pass = By.name("password");
+    public By loginButton = By.xpath("//button[contains(@class, 'bigButton__color-organish')]");
```
Comment (mkovsinv) [inline src/main/java/locators/PageLocators.java]: bigButton__color-organish replace with text

File: src/main/java/core/uttility/WebDriver.java [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+import org.openqa.selenium.chrome.ChromeOptions;
+import org.openqa.selenium.firefox.FirefoxDriver;
+import org.openqa.selenium.firefox.FirefoxOptions;
+
+import java.time.Duration;
+
+public class WebDriver {
+    private org.openqa.selenium.WebDriver driver;
+
+    public void setUp() {
```
Comment (mkovsinv) [inline src/main/java/core/uttility/WebDriver.java]: replace with constructor

File: src/main/java/core/uttility/WebDriver.java [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+                driver = new FirefoxDriver(firefoxOptions);
+            }
+            driver.manage().timeouts().implicitlyWait(Duration.ofSeconds(30));
+
+        } catch (Exception e) {
+            logError("Failed to start the WebDriver: " + e.getMessage());
+        }
+    }
+
+    private void setScreenResolution(Object options, String browserSize) {
```
Comment (mkovsinv) [inline src/main/java/core/uttility/WebDriver.java]: remove method, move options closer to driver initialization

File: src/main/java/core/uttility/TestUtilities.java [hunk 14]
```diff
@@ -0,0 +1,16 @@
+package core.uttility;
+
+import io.restassured.response.ValidatableResponse;
+import org.testng.Assert;
+
+public class TestUtilities {
+    public static void assertStatusCodeAndErrorMessage(ValidatableResponse response, int expectedStatusCode) {
+        response.assertThat().statusCode(expectedStatusCode);
```
Comment (mkovsinv) [inline src/main/java/core/uttility/TestUtilities.java]: use test ng assertions

File: src/main/java/core/uttility/TestUtilities.java [hunk 15]
```diff
@@ -0,0 +1,16 @@
...
+
+        // Additional assertion for error message if status code is not as expected
+        if (response.extract().statusCode() != expectedStatusCode) {
+            String errorMessage = response.extract().asString();
+            Assert.fail("Expected status code " + expectedStatusCode + " but got: " + response.extract().statusCode() + "\nError message: " + errorMessage);
```
Comment (mkovsinv) [inline src/main/java/core/uttility/TestUtilities.java]: fail replace with Assert(true\false\equals)

File: src/main/java/core/uttility/ApiCaller.java [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+import io.restassured.http.ContentType;
+import io.restassured.response.Response;
+import io.restassured.specification.RequestSpecification;
+
+import java.util.Map;
+
+import static io.restassured.RestAssured.given;
+
+public class ApiCaller {
+
```
Comment (mkovsinv) [inline src/main/java/core/uttility/ApiCaller.java]: authToken to class variable

File: src/main/java/core/uttility/ApiCaller.java [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+    public Response callEndpoint(String httpMethod, String endpoint, String authToken, Map<String, Object> requestBody) {
+        Response response = null;
+
+        // Configure the request using .given() and .header()
```
Comment (mkovsinv) [inline src/main/java/core/uttility/ApiCaller.java]: separated http methods

File: src/main/java/core/configuration/Configs.java (hunk 6 above)
Comment (Ania911) [reply src/main/java/core/configuration/Configs.java]: done

File: src/main/java/core/uttility/TestUtilities.java (hunk 14 above)
Comment (Ania911) [reply src/main/java/core/uttility/TestUtilities.java]: done

File: src/main/java/core/uttility/ApiCaller.java (hunk 17 above)
Comment (Ania911) [reply src/main/java/core/uttility/ApiCaller.java]: done

File: src/main/java/core/assertion/Assertion.java [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+public class Assertion {
+    private BasePage basePage;
+    public Assertion(BasePage basePage) {
+        this.basePage = basePage;
+    }
+    public void assertSuccessMessageByLocator(By successMessageLocator, String expectedMessage, String error) {
+        basePage.waitForVisibilityOfElement(successMessageLocator, 10);
+        WebElement successMessage = basePage.findElement(successMessageLocator);
+        String actualMessage = successMessage.getText();
+        basePage.assertTextEquals(actualMessage, expectedMessage, error);
```
Comment (mkovsinv) [inline src/main/java/core/assertion/Assertion.java]: actions -> get ......

File: src/main/java/core/assertion/Assertion.java [hunk 19]
```diff
@@ -0,0 +1,29 @@
...
+    }
+
+    public String isElementPresent(By locator, String successMessage, String failureMessage) {
```
Comment (mkovsinv) [inline src/main/java/core/assertion/Assertion.java]: is == true\false

File: src/main/java/core/configuration/ConfigReader.java [hunk 20]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+    public static ConfigReader getAPIInstance() {
+        if (apiInstance == null) {
+            apiInstance = new ConfigReader("api-config.yaml");
+        }
+        return apiInstance;
+    }
+
+    private void loadConfig(String configFile) {
```
Comment (mkovsinv) [inline src/main/java/core/configuration/ConfigReader.java]: move to base class

File: src/main/java/core/configuration/ConfigReader.java [hunk 21]
```diff
@@ -0,0 +1,62 @@
...
+                }
+            } else {
+                Log.error("Configuration file '" + configFile + "' not found");
+            }
+        } catch (Exception e) {
+            Log.error("An error occurred while loading the configuration '" + configFile + "': " + e.getMessage());
+        }
+    }
+
+    public Object getUIValue(String key) {
```
Comment (mkovsinv) [inline src/main/java/core/configuration/ConfigReader.java]: getValue - > override

File: src/main/java/core/pageActions/DashBoardActions.java [hunk 22]
```diff
@@ -0,0 +1,43 @@
...
+
+public class DashBoardActions extends DashboardPage {
+    public String dashboardName = generateRandomDashboardName();
+
+    public DashBoardActions(WebDriver driver) {
+        super(driver);
+    }
+
+    public void addNewDashboard() {
+        Log.info("User add new dashboard");
```
Comment (mkovsinv) [inline src/main/java/core/pageActions/DashBoardActions.java]: public String dashboardName = generateRandomDashboardName();

File: src/main/java/core/pageActions/DashBoardActions.java [hunk 23]
```diff
@@ -0,0 +1,43 @@
...
+        clickAddNewDashboard();
+        enterName(dashboardName);
+        clicksSaveDashboard();
+        System.out.println(dashboardName);
```
Comment (mkovsinv) [inline src/main/java/core/pageActions/DashBoardActions.java]: return dashboardName

File: src/main/java/core/pageActions/DashBoardActions.java [hunk 24]
```diff
@@ -0,0 +1,43 @@
...
+    }
+    public String getDashboardName() {
+        return dashboardName;
+    }
```
Comment (mkovsinv) [inline src/main/java/core/pageActions/DashBoardActions.java:21]: remove

File: src/main/java/core/pageActions/DashBoardActions.java [hunk 25]
```diff
@@ -0,0 +1,43 @@
...
+
+    public void editDashboard() {
```
Comment (mkovsinv) [inline src/main/java/core/pageActions/DashBoardActions.java]: dashboardName(String dashboardName)

File: src/main/java/core/pageActions/SettingPageActions.java [hunk 26]
```diff
@@ -0,0 +1,22 @@
...
+import org.openqa.selenium.WebDriver;
+import pages.SettingsPage;
+
+public class SettingPageActions extends SettingsPage {
+
+    public SettingPageActions(WebDriver driver) {
+        super(driver);
+    }
+
+    public void selectFromDropDown() {
```
Comment (mkovsinv) [inline src/main/java/core/pageActions/SettingPageActions.java]: String option

File: src/main/java/core/uttility/WebDriverSetUp.java [hunk 27]
```diff
@@ -0,0 +1,70 @@
...
+    public void setUp() {
+//         Get the browser configuration value from the config
+        String browser = UIConfig.getBrowser();
+        String browserSize = UIConfig.getBrowserSize();
+        try {
+//        Initialize the WebDriver based on the browser
+            if ("chrome".equalsIgnoreCase(browser)) {
+//            https://bonigarcia.dev/webdrivermanager/ - setup gecko driver
+                WebDriverManager.chromedriver().setup();
+                ChromeOptions chromeOptions = new ChromeOptions();
```
Comment (mkovsinv) [inline src/main/java/core/uttility/WebDriverSetUp.java:29]: addArguments("--start-maximized");

...
....

File: src/main/java/core/pageActions/LoginPageActions.java [hunk 28]
```diff
@@ -0,0 +1,28 @@
...
+import core.uttility.Log;
+import org.openqa.selenium.WebDriver;
+import pages.LoginPage;
+
+public class LoginPageActions extends LoginPage {
+
+    public LoginPageActions(WebDriver driver) {
+        super(driver);
+    }
+
```
Comment (mkovsinv) [inline src/main/java/core/pageActions/LoginPageActions.java:14]: @Step

**Pull request merged**